### ISSUE TITLE
Stdin build silently ignores non-inline mod items

### ISSUE THREAD

**Issue by mystor** [rust-lang/rust#65601]
Missing external modules don't report a build error when a crate is built from stdin.

**Expected**: When parsing an external submodule from stdin, a compiler error (or warning) is emitted noting that external submodules are not supported for stdio crates.

**Actual**: No diagnostics are emitted and an empty submodule is implicitly inserted.

Generally if a `mod` statement references a file which does not exist, an E0583 compile error will be emitted. However, when building a crate from stdin, no relative filesystem is present, yet no compile errors are emitted, even with explicit `#[path="x"]` attributes. Instead they are implicitly substituted with empty inline modules.

```bash
# Surprising stdin cases
$ echo 'mod x;' | rustc --crate-type=rlib -  # Builds
$ echo '#[path="/y"] mod x;' | rustc --crate-type=rlib -  # Builds

# Working file case
$ echo 'mod x;' > y.rs
$ rustc --crate-type=rlib y.rs
error[E0583]: file not found for module `x`
 --> y.rs:1:5
[...]
```

I'm not super familiar with rustc, so this might not be handy, but I'm guessing this is caused by the `Parser::recurse_into_file_modules` flag being unset when parsing from a string rather than a file ([`rustc_interface::passes::parse`](https://github.com/rust-lang/rust/blob/e5b8c118a38e8f3319813de56386bf43751582d7/src/librustc_interface/passes.rs#L67-L70) -> [`syntax::parse::parse_crate_from_source_str`](https://github.com/rust-lang/rust/blob/e5b8c118a38e8f3319813de56386bf43751582d7/src/libsyntax/parse/mod.rs#L73-L76) -> [`syntax::parse::new_parser_from_source_str`](https://github.com/rust-lang/rust/blob/e5b8c118a38e8f3319813de56386bf43751582d7/src/libsyntax/parse/mod.rs#L103-L112)), leading to a dummy module [being inserted](https://github.com/rust-lang/rust/blob/e5b8c118a38e8f3319813de56386bf43751582d7/src/libsyntax/parse/parser/module.rs#L56-L77) rather than trying to access the filesystem or emit an error during the parse.

**Comment by Enselic** [comment]
This was fixed in nightly-2020-03-20, probably by https://github.com/rust-lang/rust/pull/69838

I don't think this is worth adding a regression test for this TBH, so I think we should simply close this issue. It is also fixed in current stable (1.71).

```console
$ echo 'mod x;' | rustc +nightly-2020-03-19 --crate-type=rlib -
$ echo 'mod x;' | rustc +nightly-2020-03-20 --crate-type=rlib -
error[E0583]: file not found for module `x`
 --> <anon>:1:1
  |
1 | mod x;
  | ^^^^^^
  |
  = help: to create the module `x`, create file "x.rs"

error: aborting due to previous error

For more information about this error, try `rustc --explain E0583`.
$ echo 'mod x;' | rustc +stable --crate-type=rlib -
error[E0583]: file not found for module `x`
 --> <anon>:1:1
  |
1 | mod x;
  | ^^^^^^
  |
  = help: to create the module `x`, create file "x.rs" or "x/mod.rs"

error: aborting due to previous error

For more information about this error, try `rustc --explain E0583`.
```

**Comment by Enselic** [comment]
On a second thought the right way to close this issue is probably by writing a regression test after all. Otherwise the bug can re-appear. I was being too lazy.

@rustbot label E-needs-test
